### ISSUE TITLE
Add app.environment to payload sent to APM Server

### ISSUE THREAD

**Issue by watson** [elastic/apm-agent-nodejs#88]
To run Elastic APM on multiple environments of the same service, you'd previously normally add a post-fix to the `serviceName`, e.g: `my-service-staging` or `my-service-prod`.

In Elastic APM 6.2 we'll introduce a new `environment` property tailored specifically for this. This means that users should use the same `serviceName` (`my-service`), and instead specify the environment via a separate property whos value should be sent up to the APM Server using `service.environment` in the payload.

We need to settle on the name of the config option, but we'll probably name it either `env` or `environment`.

Example usage:

```js
require('elastic-apm-node').start({
  serviceName: 'my-service',
  env: 'production'
})
```

The associated environment variable would then be either `ELASTIC_APM_ENV` or `ELASTIC_APM_ENVIRONMENT` depending on what name we end up choosing.

Related APM Server PR: elastic/apm-server#366

**Comment by watson** [comment]
We should default its value to be the that of `process.env.NODE_ENV` (hat tip to @Qard)

**Comment by AdriVanHoudt** [comment]
If it is called `environment` in the Elastic APM I would use the same name in the config

**Comment by watson** [comment]
Another unanswered question: Should we default to a value if neither it nor `process.env.NODE_ENV` is set? Or should it then just be left out entirely?

**Comment by watson** [comment]
(just renamed `app` to `service` in the issue description to align with the new naming conventions)

**Comment by AdriVanHoudt** [comment]
> Should we default to a value if neither it nor process.env.NODE_ENV is set?

From a user perspective, leave it blank. If you have use case for it business wise or something you can default it.

**Comment by Qard** [comment]
Most popular web frameworks internally default to 'development' when NODE_ENV is not present.

see: https://github.com/expressjs/express/blob/c0136d8b48dd3526c58b2ad8666fb4b12b55116c/lib/application.js#L71

**Comment by watson** [comment]
@roncohen I think we settled on calling the config variable `environment` as opposed to `env` right?

**Comment by roncohen** [comment]
This is blocked waiting for UI support

**Comment by watson** [comment]
@roncohen what's the status on the API UI support?

**Comment by sqren** [comment]
@alvarolobato I don't think we are tracking this issue in the UI board so created an issue https://github.com/elastic/kibana/issues/21192

**Comment by alvarolobato** [comment]
Thanks @sqren

**Comment by havber** [comment]
![bilde](https://user-images.githubusercontent.com/1674488/53419633-82b40380-39da-11e9-9c7e-2c329b5d2dfa.png)

Is there any update on the status of this? With the arrival of the search bar I assume this is no longer blocked by UI.

**Comment by sqren** [comment]
Hi @havber 

You are right that the search bar enables you to filter by environment. However, you still have to write the environment filter manually. Different environments will by default be grouped together if the service name is identical which will be confusing to many. Until we can provide a better experience we recommend against relying on environment and instead using separate service names.

We have an open issue for creating dedicated environment filters in the UI that I suggest you subscribe to:
https://github.com/elastic/kibana/issues/30943.

**Comment by geoffreybergdoll** [comment]
Hi everyone,

I found a lot of discussions/threads/issues redirecting to here (maybe the only related issue still open...) so I guess the discussion is happening here.

IMO, the UI discussion is not the right debate.
As the **General APM fields** include the **service.environment** field (see [here](https://www.elastic.co/guide/en/apm/server/current/exported-fields-apm.html#_service_fields)), any agent should implement it, regardless of how (& how good) it is used within the UI.

BTW, do you know that every other agent (except .Net) implements it? [python](https://www.elastic.co/guide/en/apm/agent/python/current/configuration.html#config-environment), [ruby](https://www.elastic.co/guide/en/apm/agent/ruby/current/configuration.html#config-environment), [RUM](https://www.elastic.co/guide/en/apm/agent/js-base/current/configuration.html#environment), [go](https://www.elastic.co/guide/en/apm/agent/go/current/configuration.html#config-environment) and [java](https://www.elastic.co/guide/en/apm/agent/java/current/config-core.html#config-environment)
Troll: Do we really want NodeJs to be compared to .Net ?  ;D

As a conclusion, I kindly request to have a second look at PR like #150 and to implement this field soon (if not now) in order to match the APM Server expectations and to be future-proof on this matter, regardless of how & when the UI will evolve.

Many thanks in advance to you, who read until here and who will upvote this long (sorry) comment.

**Comment by watson** [comment]
It was an unfortunate mistake that the `environment` variable was made public in the other APM agents before it was fully supported by Elastic APM. When we discovered this mistake we updated the docs to include a warning not to use it.

So it's on purpose we haven't released support for it in Node.js. If we could do so, we'd have removed the support in the other agents. Unfortunately, this would have been a breaking change, so we just kept the warning.

However, we've been hard at work implementing support for `environment` throughout Elastic APM since and we're now ready to release it in the next version of the stack (v7.2). We'll release support for this field at the same time in this agent.
